### ISSUE TITLE
Update downloads for new schema

### ISSUE THREAD

**Issue by phraenquex** [m2ms/fragalysis-frontend#1075]
Update Duncan's 2021/22 work to accommodate new schema in fine-grained downloads.  Includes crystallography data and stuff.

Ideally, this would be fixed for the green release (Fragalysis v2), so that #1074 is unnecessary.  @tdudgeon and @kaliif to spec out.

**Comment by phraenquex** [comment]
@tdudgeon confirms it's duplicate of #1074

**Comment by tdudgeon** [comment]
@phraenquex I think you moved the wrong issue? This is the one we want to implement, and #1074 would not be needed once this one is done?

And can you confirm that there is to be no change in functionality for the fine-grain download, just re-plumbing it to work with the new UI and model.

**Comment by boriskovar-m2ms** [comment]
OK so first experiment is in my local environment

1. Press the **Prepare download** button
2. Following request goes to BE
```POST http://127.0.0.1:8080/api/download_structures/```
with payload
```
{
  "target_name": "XX01ZVNS2B",
  "proteins": "",
  "event_info": true,
  "sigmaa_info": true,
  "diff_info": true,
  "trans_matrix_info": true,
  "NAN": false,
  "mtz_info": false,
  "cif_info": false,
  "NAN2": false,
  "map_info": false,
  "single_sdf_file": true,
  "sdf_info": false,
  "pdb_info": false,
  "bound_info": true,
  "metadata_info": true,
  "smiles_info": true,
  "static_link": false,
  "file_url": ""
}
```
and response is
```
response is:
{
    "file_url": "/code/media/downloads/1ccfe05e-a944-46e3-8ab4-c14f16171b1d/XX01ZVNS2B.zip"
}
```
3. Then I click **Download** button and attached file is downloaded

[XX01ZVNS2B.zip](https://github.com/m2ms/fragalysis-frontend/files/13739496/XX01ZVNS2B.zip)

**Comment by boriskovar-m2ms** [comment]
Same steps are reproducible on tibor's stack ```https://fragalysis-tibor-default.xchem-dev.diamond.ac.uk/``` with same results.

**Comment by boriskovar-m2ms** [comment]
Same steps are also reproducible on staging ```https://fragalysis.xchem.diamond.ac.uk/``` with same results.

**Comment by phraenquex** [comment]
One bug:  it thinks "same download" is happening, when it is not the same.

Ideally:  the moment I click "prepare downloads", it should go into the "Historic downloads" dropdown

Even better:  tell me "it's preparing, clear this window", and then pop me a toaster when it's done.  (Presumably yellow release?)

**Comment by boriskovar-m2ms** [comment]
@kaliif following are relevant api calls for the case when you make no changes in the **Download Structures** UI.
1. ```POST http://127.0.0.1:8080/api/download_structures/```
with payload:
```
{
  "target_name": "XX01ZVNS2B",
  "proteins": "",
  "event_info": false,
  "sigmaa_info": false,
  "diff_info": false,
  "trans_matrix_info": false,
  "NAN": false,
  "mtz_info": false,
  "cif_info": false,
  "NAN2": false,
  "map_info": false,
  "single_sdf_file": true,
  "sdf_info": false,
  "pdb_info": false,
  "bound_info": true,
  "metadata_info": true,
  "smiles_info": true,
  "static_link": false,
  "file_url": ""
}
```
2. Api call to get size of the file ```HEAD http://127.0.0.1:8080/api/download_structures/?file_url=/code/media/downloads/1ccfe05e-a944-46e3-8ab4-c14f16171b1d/XX01ZVNS2B.zip```
and this one fails with ```404 Not Found```

**Comment by boriskovar-m2ms** [comment]
After the download from my previous comment fails then every other download will fail - basically forever (restart won;t help because I guess this info is stored in the DB) and I guess you need to fix it in the DB or like in my case just destroy my whole local stack and start from scratch with fresh and empty DB.

here is an example of the second request (same configuration like in my very first comment):
```POST http://127.0.0.1:8080/api/download_structures/```
payload:
``` 
{
  "target_name": "XX01ZVNS2B",
  "proteins": "",
  "event_info": true,
  "sigmaa_info": true,
  "diff_info": true,
  "trans_matrix_info": true,
  "NAN": false,
  "mtz_info": false,
  "cif_info": false,
  "NAN2": false,
  "map_info": false,
  "single_sdf_file": true,
  "sdf_info": false,
  "pdb_info": false,
  "bound_info": true,
  "metadata_info": true,
  "smiles_info": true,
  "static_link": false,
  "file_url": ""
}
```
response:
```208 Already Reported```
```{"message":"Zip being rebuilt - please try later"}```

**Comment by boriskovar-m2ms** [comment]
@phraenquex So the _easier_ option (just put immediately download to combo box) ended up to be **really** complicated because now the dialog is relying on pessimistic updates and for this I would have to rework the inner workings of the dialog to support optimistic updates which **always** opens can of worms and I would need to handle error states and rollback changes and other problems. Every download creates a special download tag which is kind of hackish way we decided on long time ago and it kind of obfuscates the whole process but it works so...

So I found out that actually partial implementation of #1256 is **waaaay** easier. So basically now you can close the dialog and toasts will inform you on the result of the operation (as of time of writing backend side is still broken - and if download fails then every other attempted download will return that same download is already being prepared - see my previous comment @kaliif ). After the download is ready go back to the download dialog and choose your download from the combo box. Download is added to the combo box only after given download is successfully prepared - pessimistic update and this is hard to change - see previous paragraph. In the future I can add links/buttons to the toasts so you can directly either open the download dialog with preselected download from the combo box or directly download prepared zip file. We can add it as a part of the #1256 ticket. 

There probably are some blemishes here and there but overall it should work.

**Comment by phraenquex** [comment]
Thanks @boriskovar-m2ms, good work!  Some questions:

* will the backend support this by today, or is it more complex?  (@kaliif )
* any reason NOT to re-assign #1256 to purple release?  (I've preemptively assigned it @boriskovar-m2ms )
* we certainly need the toaster to have a link to the prepared download.

Also:  will the actual bug be fixed today, that triggered yesterday's discussion?

**Comment by kaliif** [comment]
I'll look into it, but I'd say unlikely to happen today, I still haven't gotten to the root cause of this.

**Comment by phraenquex** [comment]
Thanks, helpful.  But we'll leave it on the purple release.

**Comment by boriskovar-m2ms** [comment]
Remaining frontend work is moved to #1256.

**Comment by kaliif** [comment]
I just created a pull request (https://github.com/xchem/fragalysis-backend/pull/482) that should fix this issue on the backend side. About the frontend the payload needs some adjustment:
> {
>   "target_name": "XX01ZVNS2B",
>   "proteins": "",
>   "event_info": true,
>   "sigmaa_info": true,
>   "diff_info": true,
>   "trans_matrix_info": true,
>   "NAN": false,
>   "mtz_info": false,
>   "cif_info": false,
>   "NAN2": false,
>   "map_info": false,
>   "single_sdf_file": true,
>   "sdf_info": false,
>   "pdb_info": false,
>   "bound_info": true,
>   "metadata_info": true,
>   "smiles_info": true,
>   "static_link": false,
>   "file_url": ""
> }
> ```

If you look at the endpoint (https://fragalysis.xchem.diamond.ac.uk/api/download_structures/), the schema change brought some changes to the DRF serializer and therefore to the endpoint. The following changes are necessary:
- trans_matrix_info isn't captured anymore so that's not available
- bound_info -> bound_file
- event_info -> event_file
- pdb_info -> apo_file
- sigmaa_info -> xmap_2fofc_file
- diff_info -> xmap_fofc_file


The last two seem to be called correctly in the download modal but not in the payload.

**Comment by phraenquex** [comment]
Thanks @kaliif.  

Please confirm why the last two are called 
* xmap_2fofc and not xmap_sigmaa
* xmap_fofc and not xmap_diff

In #1166, I'd replied to your question to clarify that any new code should use "sigmaa" and "diff" (rather than 2fofc and fofc).  At least, it's what I tried to say.  I now reiterate that.

About `trans_matrix_info`:  that should return the file generated by @ConorFWild's XChemAlign code, where all the transformation matrices are captured.  If that is not currently being ingested by the target loader, then that is a bug - ticket #1055 (I think).

**Comment by kaliif** [comment]
@phraenquex sorry, missed that bit about renaming the files, will change it.

According to the spreadsheet (https://docs.google.com/spreadsheets/d/10h2IfxstgWnnEWe0IB-q0SduEcDAU35RKZkzu6R7fjs/edit#gid=0), `trans_matrix_info` field is removed. This info doesn't seem to be present in XCA output either. Is this incorrect?

**Comment by phraenquex** [comment]
Ah - I don't remember seeing (or rather, looking at) that spreadsheet.  

Certainly trans_matrix was a placeholder in the earlier implementation.  It's considerably more complex now, after @ConorFWild thought hard about it - but it's still a Thing, I believe.

@ConorFWild can you point @kaliif to the output you generate that documents all the transformations you discover?  

If you're not sure what I mean:  I remember asking you whether the transformations can't just calculated on the fly, from the residue ranges;  and you replied emphatically that no, it was worth writing out, because the calculations are non-trivial.

**Comment by phraenquex** [comment]
@kaliif please also update that spreadsheet with the name changes.

**Comment by phraenquex** [comment]
@boriskovar-m2ms please respond to my questions from 3 weeks ago: 

> Thanks @boriskovar-m2ms, good work! Some questions:
>    * will the backend support this by today, or is it more complex? (@kaliif )
>    * any reason NOT to re-assign https://github.com/m2ms/fragalysis-frontend/issues/1256 to purple release? (I've preemptively assigned it @boriskovar-m2ms )
>    * we certainly need the toaster to have a link to the prepared download.
>
> Also: will the actual bug be fixed today, that triggered yesterday's discussion?

**Comment by ConorFWild** [comment]
The relevant files are:

neighbourhood_transforms.yaml --> Transforms that align ligand neighbourhoods to one another
conformer_site_transforms.yaml --> Transforms that align conformer sites to their canonical site
reference_structure_transforms.yaml --> Transforms that align canonical sites to a reference structure

**Comment by phraenquex** [comment]
@ConorFWild do you suggest the download returns them as separate files?

If yes:  @kaliif feel free to call the database fields as you see fit, but return them all if that one radio button is checked in the frontend modal.  
Maybe this?
* transforms_neighbourhood,
* transforms_conformer_site
* transforms_reference_structure

@ConorFWild:  you also need to update the documentation that's put into the self-documenting PDF that's added to each zip archive.  @alanbchristie ought to be able to point you to the relevant code, but you'll have to write the docs.  

(This last thing is optional for purple release, though I suspect only slightly optional, given how worried everyone is (correctly!) about documentation lagging release.)

**Comment by kaliif** [comment]
Found the files. Which models would be the most sensible to attach them to?

**Comment by boriskovar-m2ms** [comment]
> @boriskovar-m2ms please respond to my questions from 3 weeks ago:
> 
> > Thanks @boriskovar-m2ms, good work! Some questions:
> > 
> > * will the backend support this by today, or is it more complex? (@kaliif )
> > * any reason NOT to re-assign [Streamline the download-prepare workflow #1256](https://github.com/m2ms/fragalysis-frontend/issues/1256) to purple release? (I've preemptively assigned it @boriskovar-m2ms )
> > * we certainly need the toaster to have a link to the prepared download.
> > 
> > Also: will the actual bug be fixed today, that triggered yesterday's discussion?

* I think it's not controversial to assign it to purple release
* added link to toasts before Christmas

**Comment by phraenquex** [comment]
@kaliif those transformations should be added to v_experiment_uploads.  And for the download, take the only the entries for the latest download.  (@ConorFWild please confirm that these transformations are always relevant to all models, even those uploaded previously.)

**Comment by kaliif** [comment]
> @alanbchristie will look into large download file (@boriskovar-m2ms reports starting with 1GB and download gets 7GB file)

I just tried to gzip 7GB file and it went to 250MB. Looks like a compression issue.

**Comment by phraenquex** [comment]
Waiting for #1284 to test.

**Comment by alanbchristie** [comment]
I suspect the 60-second timeout has something to do with stanbard **keep-alive** timeouts, although I have yet to confirm this.

**workaround**
If you trigger a download and it fails after 60 seconds you could just swait a minute or two and try it again. If the construction of the download is successful you will get access to it on the next query - the constructed download file is kept for a significant period of time (90 minutes on the Tibor stack).

(Tibor's stack contains general fixes to the Download process btw).

**Comment by kaliif** [comment]
> @kaliif those transformations should be added to v_experiment_uploads. And for the download, take the only the entries for the latest download. (@ConorFWild please confirm that these transformations are always relevant to all models, even those uploaded previously.)

Both upload and download now handle trans matrices (merged to staging branch, not deployed to stack yet).

**Comment by tdudgeon** [comment]
@phraenquex @Waztom "Downloads need to be tested after target loader is run".

I think we need to define what needs to be tested.
Sometimes it is obvious. X has changed so X needs to be tested.
But more generally, there is an almost infinite combination of options to set/unset which means it would take ages to test everything.
So if this is to be routinely tested I think we need a written "manual test spec" that needs to be executed.
And maybe 2 specs (basic tests and comprehensive tests).

Potentially this could be unit tested with something like cucumber, but with everything changing so quickly that's not going to be possible until we get some level of stability, so for now I think we must have a manual test plan for this.
